node/core/backing/src/lib.rs

```diff
@@ -786,626 +741,1279 @@ async fn validate_and_make_available(
-			// create a span only for candidates we're already aware of.
-			self.get_unbacked_statement_child(
-				root_span,
-				candidate_hash,
-				statement.validator_index(),
-			)
-		};
+				return SecondingAllowed::No
+			},
+			Ok((depths, head, leaf_state)) => {
+				for depth in &depths {
+					if leaf_state.seconded_at_depth.contains_key(&depth) {
+						gum::debug!(
+							target: LOG_TARGET,
+							?candidate_hash,
+							depth,
+							leaf_hash = ?head,
+							"Refusing to second candidate at depth - already occupied."
+						);
 
-		if let Err(ValidatorIndexOutOfBounds) = self
-			.dispatch_new_statement_to_dispute_coordinator(ctx.sender(), candidate_hash, &statement)
-			.await
-		{
-			gum::warn!(
-				target: LOG_TARGET,
-				session_index = ?self.session_index,
-				relay_parent = ?self.parent,
-				validator_index = statement.validator_index().0,
-				"Supposedly 'Signed' statement has validator index out of bounds."
-			);
+						return SecondingAllowed::No
+					}
+				}
 
-			return Ok(None)
+				membership.push((*head, depths));
+			},
 		}
+	}
 
-		let stmt = primitive_statement_to_table(statement);
+	// At this point we've checked the depths of the candidate against all active
+	// leaves.
+	SecondingAllowed::Yes(membership)
+}
 
-		let summary = self.table.import_statement(&self.table_context, stmt);
+#[overseer::contextbounds(CandidateBacking, prefix = self::overseer)]
+async fn handle_validated_candidate_command<Context>(
+	ctx: &mut Context,
+	state: &mut State,
+	relay_parent: Hash,
+	command: ValidatedCandidateCommand,
+	metrics: &Metrics,
+) -> Result<(), Error> {
+	match state.per_relay_parent.get_mut(&relay_parent) {
+		Some(rp_state) => {
+			let candidate_hash = command.candidate_hash();
+			rp_state.awaiting_validation.remove(&candidate_hash);
+
+			match command {
+				ValidatedCandidateCommand::Second(res) => match res {
+					Ok(outputs) => {
+						let BackgroundValidationOutputs {
+							candidate,
+							commitments,
+							persisted_validation_data,
+						} = outputs;
+
+						if rp_state.issued_statements.contains(&candidate_hash) {
+							return Ok(())
+						}
 
-		let unbacked_span = if let Some(attested) = summary
-			.as_ref()
-			.and_then(|s| self.table.attested_candidate(&s.candidate, &self.table_context))
-		{
-			let candidate_hash = attested.candidate.hash();
-			// `HashSet::insert` returns true if the thing wasn't in there already.
-			if self.backed.insert(candidate_hash) {
-				let span = self.remove_unbacked_span(&candidate_hash);
+						// sanity check that we're allowed to second the candidate
+						// and that it doesn't conflict with other candidates we've
+						// seconded.
+						let fragment_tree_membership = match seconding_sanity_check(
+							ctx,
+							&state.per_leaf,
+							&state.implicit_view,
+							candidate_hash,
+							candidate.descriptor().para_id,
+							persisted_validation_data.parent_head.hash(),
+							commitments.head_data.hash(),
```

Comment:
Doesn't this head data correspond to the result of candidate's execution?
I assume we need a parent head here which comes from the PVD.

Reply:
the parent-head hash is passed just above and it seems the head-data hash here is used only to detect & remove cycles. seems correct to me.